package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,371 @@
+        self._metric = metric
+        self._cutoff = cutoff
+        self._weights = weights
+        self._calculated = False
+        # remember that this must be called before referencing self.nframes
+        self._setup_frames(traj, start, stop, step)
+
+    def _prepare(self):
+        self.dist_matrix = np.zeros((self.nframes, self.nframes))
+
+    def _single_frame(self):
+        iframe = self._ts.frame
+        i_ref = self.atoms.positions
+        # diagonal entries need not be calculated due to metric(x,x) == 0 in
+        # theory, _ts not updated properly. Possible savings by setting a
+        # cutoff for significant decimal places to sparsify matrix
+        for j, ts in enumerate(self._u.trajectory[iframe:self.stop:self.step]):
+            self._ts = ts
+            j_ref = self.atoms.positions
+            dist = self._metric(i_ref, j_ref, weights=self._weights)
+            self.dist_matrix[self._frame_index, j+self._frame_index] = (
+                dist if dist > self._cutoff else 0)
+            self.dist_matrix[j+self._frame_index, self._frame_index] = (
+                self.dist_matrix[self._frame_index, j+self._frame_index])
+        self._ts = self._u.trajectory[iframe]
+
+    def _conclude(self):
+        self._calculated = True
+
+    def save(self, filename):
+        np.save(filename, self.dist_matrix)
+        logger.info("Wrote the distance-squared matrix to file %r", filename)
+
+
+class DiffusionMap(object):
+    """Non-linear dimension reduction method
+
+    Dimension reduction with diffusion mapping of selected structures in a
+    trajectory.
+
+    Attributes
+    ----------
+    eigenvalues: array
+        Eigenvalues of the diffusion map
+    eigenvectors: array
+        Eigenvectors of the diffusion map
+    diffusion_space : array
+        After calling `transform(num_eigenvectors)` the diffusion map embedding
+        into the lower dimensional diffusion space will exist here.
+
+    Methods
+    -------
+    run()
+        Constructs an anisotropic diffusion kernel and performs eigenvalue
+        decomposition on it.
+    transform(num_eigenvectors)
+        Perform an embedding of a frame into the eigenvectors representing
+        the collective coordinates.
+    """
+
+    def __init__(self, u, epsilon=1, manifold_density=None,
+                 timescale=1, **kwargs):
+        """
+        Parameters
+        -------------
+        u : MDAnalysis Universe or DistanceMatrix object
+            Can be a Universe, in which case one must supply kwargs for the
+            initialization of a DistanceMatrix. Otherwise, this can be a
+            DistanceMatrix already initialized. Either way, this will be made
+            into a diffusion kernel.
+        epsilon : Float
+            Specifies the method used for the choice of scale parameter in the
+            diffusion map. More information in [1], [2] and [3], Default: 1.
+        manifold_density: list, optional
```

Comment:
You're absolutely right, I will do some research and get back to you with a
satisfying explanation or eliminate it as a parameter.
On Fri, Jul 1, 2016 at 1:18 PM kain88-de <notifications@github.com> wrote:

> In package/MDAnalysis/analysis/diffusionmap.py
> <https://github.com/MDAnalysis/mdanalysis/pull/863#discussion_r69350871>:
>
> > +    """
> > +
> > +    def __init__(self, u, epsilon=1, manifold_density=None,
> > +                 timescale=1, **kwargs):
> > +        """
> > +        Parameters
> > +        -------------
> > +        u : MDAnalysis Universe or DistanceMatrix object
> > +            Can be a Universe, in which case one must supply kwargs for the
> > +            initialization of a DistanceMatrix. Otherwise, this can be a
> > +            DistanceMatrix already initialized. Either way, this will be made
> > +            into a diffusion kernel.
> > +        epsilon : Float
> > +            Specifies the method used for the choice of scale parameter in the
> > +            diffusion map. More information in [1], [2] and [3], Default: 1.
> > +        manifold_density: list, optional
>
> Yeah Focker-Plank equation and diffusion are related I know (also good the
> correspondence can be shown for diffusion maps). Yes MD-simulations can be
> modeled with Fokker-Plank equations. What I would like to know is *what
> is a manifold_density?* Your answer doesn't give me the description. I
> still don't know what this parameter does from reading the docs nor do I
> know what sensible values are as a user nor could I guess them. Please
> explain what the manifold_density is for the diffusion maps and how they
> influence the result. If we can't explain it think about leaving it out.
>
> Please don't link to blog posts. We don't know how long they will be
> around and MDAnalysis will likely be around for another 10 years so I
> rather keep everything in the docs self contained. References to
> publications are OK.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/863/files/b6c309a5dac5b2a5fc816fb0c2eccfd1a4a81b93#r69350871>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARv5Jw8X2-0Ck7TR5MuS9Tcw5abKjks5qRXYagaJpZM4IpYIS>
> .
>
-- 
Have a wonderful day,
John J. Detlefs